jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/internal/WebSocketChannelState.java

```diff
@@ -19,140 +19,223 @@
+    {
+        synchronized (this)
+        {
+            _outgoingSequence = getNextState(frame, _outgoingSequence);
+
+            if (frame.getOpCode() == OpCode.CLOSE)
             {
-                State closed = new State("CLOSED", false, false, closeStatus);
-                if (state.compareAndSet(s, closed))
-                    return true;
+                _closeStatus = CloseStatus.getCloseStatus(frame);
+                _channelState = updateChannelState(_channelState, _incomingSequence, _outgoingSequence);
             }
+
+            return _channelState.equals(State.CLOSED);
         }
     }
 
-    public boolean onCloseOut(CloseStatus closeStatus)
+
+
+    public boolean checkIncoming(Frame frame) throws ProtocolException
     {
-        while (true)
+        synchronized (this)
         {
-            State s = state.get();
-
-            if (!s.outOpen)
-                throw new IllegalStateException(state.get().toString());
+            _incomingSequence = getNextState(frame, _incomingSequence);
 
-            if (s.inOpen)
-            {
-                State closedOut = new State("OCLOSED", true, false, closeStatus);
-                if (state.compareAndSet(s, closedOut))
-                    return false;
-            }
-            else
+            if (frame.getOpCode() == OpCode.CLOSE)
             {
-                State closed = new State("CLOSED", false, false, closeStatus);
-                if (state.compareAndSet(s, closed))
-                    return true;
+                _closeStatus = CloseStatus.getCloseStatus(frame);
+                _channelState = updateChannelState(_channelState, _incomingSequence, _outgoingSequence);
             }
+
+            return _channelState.equals(State.CLOSED);
         }
     }
 
-    public boolean onClosed(CloseStatus closeStatus)
+
+
+    private static State updateChannelState(State state, byte _incomingSequence, byte _outgoingSequence)
     {
-        while (true)
+        switch (state)
         {
-            State s = state.get();
-            if (!s.outOpen && !s.inOpen)
-                return false;
+            case OPEN:
+            case ICLOSED:
+            case OCLOSED:
+                break;
+            default:
+                throw new IllegalStateException(state.toString());
+        }
 
-            State newState = new State("CLOSED", false, false, closeStatus);
-            if (state.compareAndSet(s, newState))
-                return true;
+        State newState = state;
+        if ((_outgoingSequence == OpCode.CLOSE) && (_incomingSequence == OpCode.CLOSE))
+            newState = State.CLOSED;
+        else if (_outgoingSequence == OpCode.CLOSE)
+            newState = State.OCLOSED;
+        else if (_incomingSequence == OpCode.CLOSE)
+            newState = State.ICLOSED;
+
+        return newState;
+    }
+
+
+    public static byte getNextState(Frame frame, byte state) throws ProtocolException
```

Comment:
pass in the opcode rather than the frame

Reply:
Also `state` is the wrong name.  It should be `sequence` and `getNextSequence`.

I also think the real `state` should be passed in (see below):